crates/matrix-sdk/src/authentication/service.rs

```diff
@@ -0,0 +1,529 @@
+    /// the service has been configured with a homeserver.
+    #[error("A successful call to configure_homeserver must be made first.")]
+    ClientMissing,
+    /// The supplied string couldn't be parsed as either a server name or a
+    /// homeserver URL.
+    #[error("Failed to construct a server name or homeserver URL: {0}")]
+    InvalidServerName(IdParseError),
+    /// There is no server responding at the supplied URL.
+    #[error("Failed to discover a server")]
+    ServerNotFound,
+    /// There is a server at the supplied URL, but it is neither a homeserver
+    /// nor is it hosting a well-known file for one.
+    #[error("Failed to discover a homeserver")]
+    HomeserverNotFound,
+    /// The discovered well-known file is invalid.
+    #[error("Failed to parse the well-known file: {0}")]
+    InvalidWellKnownFile(DeserializationError),
+    /// Sliding sync is required, but isn't configured in the homeserver's
+    /// well-known file.
+    #[error("The homeserver doesn't provide a trusted sliding sync proxy in its well-known configuration.")]
+    SlidingSyncNotAvailable,
+    /// An error occurred whilst trying to use the supplied base path.
+    #[error("Failed to use the supplied base path.")]
+    InvalidBasePath,
+    /// An unknown error occurred.
+    #[error("An error occurred: {0}")]
+    Generic(Box<dyn Error>),
+}
+
+/// Details about a homeserver's login capabilities.
+#[derive(Clone, Debug)]
+pub struct HomeserverLoginDetails {
+    /// The URL of the homeserver.
+    pub url: Url,
+    /// Whether the homeserver supports login using OIDC as defined by MSC3861.
+    #[cfg(feature = "experimental-oidc")]
+    pub supports_oidc_login: bool,
+    /// Whether the homeserver supports the password login flow.
+    pub supports_password_login: bool,
+}
+
+impl AuthenticationService {
+    /// Creates a new service to authenticate a user with.
+    pub fn new(
+        user_agent: Option<String>,
+        #[cfg(feature = "experimental-sliding-sync")] custom_sliding_sync_proxy: Option<String>,
+    ) -> Self {
+        AuthenticationService {
+            user_agent,
+            client: RwLock::new(None),
+            homeserver_details: RwLock::new(None),
+            #[cfg(feature = "experimental-sliding-sync")]
+            custom_sliding_sync_proxy: RwLock::new(custom_sliding_sync_proxy),
+        }
+    }
+
+    /// Returns the homeserver details for the currently configured homeserver,
+    /// or `None` if a successful call to `configure_homeserver` is yet to be
+    /// made.
+    pub fn homeserver_details(&self) -> Option<HomeserverLoginDetails> {
+        self.homeserver_details.read().unwrap().clone()
+    }
+
+    /// Updates the service to authenticate with the homeserver for the
+    /// specified address.
+    pub async fn configure_homeserver(
+        &self,
+        server_name_or_homeserver_url: String,
+    ) -> Result<(), AuthenticationError> {
```

Comment:
Maybe this service thing is all pointless. This entire fn could just be part of the client builder?

Reply:
Yes indeed, we figured the responsibilities of the authentication service, and they're threefold:

- try to guess, based on a URL, if it should be treated as a homeserver URL or an arbitrary URL, by doing the well-known discovery dance, and use that -> this can probably go into the `ClientBuilder`
- perform all the OIDC login steps -> this could go as a helper method in `Oidc`
- transition from a client with no user-id to one that has a user-id -> this needs a bit more thought in general

Reply:
Following this conversation I've opened #3129 for the bullet point here.